js/private/yaml.bzl

```diff
@@ -0,0 +1,413 @@
+        "indent": indent if not flow else None,
+        "flow": flow,
+    }
+
+def _new_SEQUENCE(index, flow, indent = None):
+    return {
+        "id": _STATE.SEQUENCE,
+        "indent": indent if not flow else None,
+        "index": index,
+        "flow": flow,
+    }
+
+def _normalize_yaml(yaml):
+    yaml = yaml.replace("\r", "")
+    if not yaml.endswith("\n"):
+        yaml = yaml + "\n"
+    return yaml
+
+def _initialize_result_value(starlark, stack):
+    "Initialize empty starlark maps or list values for the current pseudostates in the stack"
+    kns_states = _get_key_and_sequence_states(stack)
+
+    if len(kns_states) == 0:
+        return
+    else:
+        if starlark["result"] == None:
+            starlark["result"] = _empty_value_for_state(kns_states[0])
+        curr_result = starlark["result"]
+        for (i, state) in enumerate(kns_states[0:-1]):
+            if type(curr_result) == "dict":
+                curr_result = curr_result.setdefault(state["key"], _empty_value_for_state(kns_states[i + 1]))
+            else:
+                if state["index"] >= len(curr_result):
+                    curr_result.append(_empty_value_for_state(kns_states[i + 1]))
+                curr_result = curr_result[state["index"]]
+
+def _set_result_value(starlark, stack, value):
+    "Add a new value to the starlark result corresponding to the last pseudostate in the stack"
+    kns_states = _get_key_and_sequence_states(stack)
+    if len(kns_states) == 0:
+        starlark["result"] = value
+    else:
+        curr_result = starlark["result"]
+        for state in kns_states[0:-1]:
+            if type(curr_result) == "dict":
+                curr_result = curr_result[state["key"]]
+            else:
+                curr_result = curr_result[state["index"]]
+        if type(curr_result) == "dict":
+            curr_result[kns_states[-1]["key"]] = value
+        else:
+            curr_result.append(value)
+
+def _empty_value_for_state(state):
+    if state["id"] == _STATE.KEY:
+        return {}
+    elif state["id"] == _STATE.SEQUENCE:
+        return []
+    else:
+        fail("State %s has no empty type" % state["id"])
+
+def _peek(stack):
+    return stack[-1]
+
+def _parse_scalar(value):
+    value = value.strip()
+    if _is_int(value):
+        return int(value)
+    elif _is_float(value):
+        return float(value)
+    elif _is_bool(value):
+        return _to_bool(value)
+    elif value.startswith("'"):
+        return value.strip("'")
+    else:
+        return value.strip("\"")
+
+def _is_float(value):
+    return value.replace(".", "", 1).isdigit()
+
+def _is_int(value):
+    return value.isdigit()
+
+def _is_bool(value):
+    return value == "true" or value == "false"
+
+def _to_bool(value):
+    if value == "true":
+        return True
+    elif value == "false":
+        return False
+    fail("Cannot convert scalar %s to a starlark boolean" % value)
+
+def _parse_key(key):
+    if key.startswith("'"):
```

Comment:
I guess we've never seen a specimen of pnpm-lock where there's also some escaped quoting within the string value?

Reply:
I don't think so, but this should only remove surrounding quotations. I'm not sure if:
```
'foo: bar
```
is valid yaml, but maybe I should do a `startswith` and `endswith` check before stripping the quotes.

Reply:
The online parsers I checked seem to puke when a key starts with a quote but allow a key to end with a quote.